sam_view.c

```diff
@@ -347,6 +421,60 @@ int main_samview(int argc, char *argv[])
                 ret = 1;
                 goto view_end;
             }
+            break;
+        case 'd':
+            if (strlen(optarg) < 4 || optarg[2] != ':') {
+                print_error_errno("view", "Invalid \"tag:value\" option: \"%s\"", optarg);
+                ret = 1;
+                goto view_end;
+            }
+
+            if (settings.tag) {
+                if (settings.tag[0] != optarg[0] || settings.tag[1] != optarg[1]) {
+                    print_error("view", "Different tag \"%s\" was specified before: \"%s\"", settings.tag, optarg);
+                    ret = 1;
+                    goto view_end;
+                }
+            } else {
+                if (!(settings.tag = calloc(3, 1))) {
+                    print_error("view", "Could not allocate memory for tag: \"%s\"", optarg);
+                    ret = 1;
+                    goto view_end;
+                }
+                memcpy(settings.tag, optarg, 2);
+            }
+
+            if (add_tag_value_single("view", &settings, optarg+3) != 0) {
+                ret = 1;
+                goto view_end;
+            }
+            break;
+        case 'D':
+            if (strlen(optarg) < 4 || optarg[2] != ':') {
```

Comment:
> 
> 
> @valeriuo I would prefer to not have the flag in the file as it makes generating the file a bit more cumbersome (else it is just a list of identifiers). It also wouldn't solve the problem for the `-d STR:STR` option.
> 
> What about `@` as delimiter?
> `-d STR@STR` `-D STR@STR`

Alternatively, you could check the delimiter for semi-colon value, i.e.
`if (strlen(optarg) < 4 || (optarg[2] != ':' && optarg[2] != ';'))`

Reply:
@valeriuo Checking for `;` too if fine for me. Is it OK if the documentation only mentions `:` as delimiter, or do you want `;` as explicit delimiter (the whole argument would need to be quoted on Unix like systems in that case).

Reply:
Only the colon should be advertised as a delimiter. The check is internal and should be accompanied by a relevant comment. Also, it's not needed for the single barcode case (`-d`), as path translation only happens if the first colon delimiter is followed by one of: slash (**/**), dot (**.**) or colon (**:**).

Reply:
I added it for both options, just in case people have specal start characters for values for their tags. I also added a test to see if `;` works as delimiter.

Reply:
Testing for `-d BC;TGCA` seems to fail on MinGW, so should the test be removed or only executed on Unix?
> 91: Filter -d BC;TGCA -D BC;C:/msys64/tmp/rtI3AvH1Ni/view.001.fobc (SAM input)samtools view: failed to open "C;C:\msys64\msys64\tmp\rtI3AvH1Ni\view.001.fobc" for reading: Invalid argument
	Failed command:
	C:/projects/samtools/samtools view -o C:/msys64/tmp/rtI3AvH1Ni/view.test091.sam -h -d BC;TGCA -D BC;C:/msys64/tmp/rtI3AvH1Ni/view.001.fobc C:/msys64/tmp/rtI3AvH1Ni/view.001.sam
.. failed ...

Reply:
And it should rightly do so, as the semi-colon is not supported as a delimiter in the command arguments.

What happens is that the string `BC;C:/msys64/tmp/rtI3AvH1Ni/view.001.fobc` gets translated into `BC;C;C:\msys64\msys64\tmp\rtI3AvH1Ni\view.001.fobc`. After you skip the tag and the delimiter (`BC;`), you are left with an invalid filename `C;C:\msys64\msys64\tmp\rtI3AvH1Ni\view.001.fobc`.

It works on Unix, because the generated paths are POSIX standard and there is no translation involved. You can discard the test.

Reply:
@valeriuo Fixed.